Add client-side models and interfaces required for multiplayer

peppy opened the pull request:
These are shared models and interfaces that are required by the SignalR server (and future client) implementations to allow communication.

Open to discussion on the namespace (and probably force push to synchronise with the client implementation).

peppy committed:
Add basic structure for multiplayer state and model components

peppy committed:
Add simple user state class

peppy committed:
Add basic server implementation

peppy committed:
Add models for users and rooms

peppy committed:
Ensure room is locked when mutating users

peppy committed:
Add locking on join/leave operations

peppy committed:
Add callbacks for join/leave events to notify other room occupants

peppy committed:
Fix serialization failure due to missing set

peppy committed:
Add room settings related model and event flow

peppy committed:
Make MultiplayerRoomUser equatable

peppy committed:
Remove unnecessary room id from leave room request

peppy committed:
Move exceptions to common code

peppy committed:
Rename multiplayer server file to match class

peppy committed:
Add thread safety for external operations on MultiplayerRoom

peppy committed:
Rename idle state

peppy committed:
Add xmldoc coverage of MultiplayerRoom

peppy committed:
Move business logic out of MultiplayerRoom

peppy committed:
Make remaining model classes nullable and serializable

peppy committed:
Return room model when joining

peppy committed:
Add missing methods to server interface

peppy committed:
Ensure multiplayer rooms are instantiated with a room ID

peppy committed:
Add client method for notifying about host changes

peppy committed:
Add notification flow for user state changes in room

peppy committed:
Mark PerformUpdate as an instant handle method (doesn't really help with anything)

peppy committed:
Add loading states

peppy committed:
Add missing states and xmldoc for all states' purposes

peppy committed:
Fix client naming and xmldoc

peppy committed:
Document room states and remove unnecessary WaitingForResults state

peppy committed:
Add methods covering match start / end

peppy committed:
Add exception to be thrown when an operation is requested requiring host when not host

peppy committed:
Add extra xmldoc covering the fact that MatchStarted is received by all users

peppy committed:
Add note about LoadRequested only being sent to a subset of room users

peppy committed:
Change locking mechanism for multiplayer rooms to use using-disposal pattern

Was required to lock over `await` calls server-side.

peppy committed:
Add missing final newline in file

bdach commented on osu.Game/Online/RealtimeMultiplayer/IMultiplayerClient.cs:
```diff
@@ -0,0 +1,65 @@
+        Task LoadRequested();
+
+        /// <summary>
+        /// Signals that a match has started. All user in the <see cref="MultiplayerUserState.Loaded"/> state should begin gameplay as soon as possible.
```
```suggestion
        /// Signals that a match has started. All users in the <see cref="MultiplayerUserState.Loaded"/> state should begin gameplay as soon as possible.
```

bdach commented on osu.Game/Online/RealtimeMultiplayer/IMultiplayerServer.cs:
```diff
@@ -0,0 +1,48 @@
+        /// As the host, update the settings of the currently joined room.
+        /// </summary>
+        /// <param name="settings">The new settings to apply.</param>
+        Task ChangeSettings(MultiplayerRoomSettings settings);
```
I imagine this could also potentially throw `NotHostException`.

bdach commented on osu.Game/Online/RealtimeMultiplayer/InvalidStateChange.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace osu.Game.Online.RealtimeMultiplayer
+{
+    public class InvalidStateChange : Exception
```
Exception class naming in this PR is sort of half and half - there are:

* `InvalidStateChange`
* `InvalidStateException`
* `NotHostException`
* `NotJoinedRoomException`
* `UserAlreadyInMultiplayerRoom`

Would probably unify by doing `InvalidStateChange` -> `InvalidStateChangeException` and `UserAlreadyInMultiplayerRoom` -> `AlreadyInRoomException`.

bdach commented on osu.Game/Online/RealtimeMultiplayer/MultiplayerRoomSettings.cs:
```diff
@@ -0,0 +1,30 @@
+
+        public override string ToString() => $"Beatmap:{BeatmapID} Mods:{string.Join(',', Mods)} Ruleset:{RulesetID}";
+
+        public static MultiplayerRoomSettings Empty() => new MultiplayerRoomSettings();
```
A bit curious why this is a thing. I'd probably either remove this or hide the parameterless ctor by making it `private`, so that there's no ambiguity whether the ctor and this static factory method differ in any capacity.

peppy replied:
parameterless needs to be there for serialization i belied. will just remove this since it doesn't add much clarity.

bdach commented on osu.Game/Online/RealtimeMultiplayer/MultiplayerRoomSettings.cs:
```diff
@@ -0,0 +1,30 @@
+        [NotNull]
+        public IEnumerable<APIMod> Mods { get; set; } = Enumerable.Empty<APIMod>();
+
+        public bool Equals(MultiplayerRoomSettings other) => BeatmapID == other.BeatmapID && Mods.SequenceEqual(other.Mods) && RulesetID == other.RulesetID;
```
Going to have to be careful about order of `Mods` with this comparer. Maybe the setter of `Mods` should ensure sorting?

peppy replied:
Hmm... I think it's safe enough for now as worst-case it's just going to trigger an update of settings to the same settings. I copied this from the existing implementation in `SpectatorState` for what that's worth.

bdach commented on osu.Game/Online/RealtimeMultiplayer/MultiplayerRoomUser.cs:
```diff
@@ -0,0 +1,42 @@
+
+        public User? User { get; set; }
+
+        public MultiplayerRoomUser(in int userId)
```
Incoming param is `int` but local field is `long` - not sure if intended.

peppy replied:
`int` is intended - i don't think we're going to hit `long` user IDs for quite a while longer.

bdach reviewed (commented):
Super nitpicky review but there's not much to go on here. Haven't considered any bigger implications of the API design or read the server-side PR, just read the structures as they are. I'd probably miss stuff - I imagine kinks in this design can be pointed out as progress goes along.

peppy committed:
Fix typo in IMultiplayerClient xmldoc

peppy committed:
Add missing exceptions to server xmldoc

peppy committed:
Standardise exception naming

peppy committed:
Change user id type to int

peppy committed:
Remove Empty() implementation for RoomSettings until otherwise necessary

peppy committed:
Add missing licence headers

peppy committed:
Change get-only property for now

peppy committed:
Change exceptions which should be returned to the user to HubException type

See
https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.signalr.hubexception?view=aspnetcore-5.0.

peppy committed:
Add ToString implementation to MultiplayerRoom for easier debug

peppy committed:
Move client state class to server implementation

Never used by client.

peppy committed:
Add JsonConstructor specs to allow for correct deserialization of readonly fields

peppy committed:
Also make InvalidStateException serializable

peppy committed:
Remove unnecessary exception type

peppy committed:
Update outdated exception references in xmldoc

peppy committed:
Add stateful client interface

bdach reviewed (dismissed):
Throwing this a tentative approve as I can't see any issues with the models myself at this point, and maybe an approval will be useful to get this merged and continue work over at the spectator server repository :+1:

Namespace is fine I think. I guess something like `osu.Game.Multiplayer.Realtime` might also work, but not sure if it's a big deal.

peppy committed:
Add room name to settings

peppy committed:
Add realtime to room categories

peppy committed:
Make room setting's BeatmapID non-nullable

peppy committed:
Change locking method to better allow cross-thread locking

peppy committed:
Make RulesetID non-nullable

smoogipoo committed:
Split server interfaces

smoogipoo committed:
Remove IStatefulMultiplayerClient

smoogipoo commented:
Removed `IStatefulMultiplayerClient` - interface proved to be useless with my mp changes.

Also split out interfaces so that my client doesn't have to implement `JoinRoom(long roomId)`. My hierarchy becomes something like the following now:
```
class StatefulMultiplayerClient : IMultiplayerClient, IMultiplayerRoomServer
{
    public Task<MultiplayerRoom> JoinRoom(Room room) // Note Room object, NOT long roomId
    {
        return Room = await connection.InvokeAsync<MultiplayerRoom>(nameof(IMultiplayerLoungeServer.JoinRoom), room.RoomId.Value);
    }
}
```
(as discussed)

peppy commented:
Seems fine.
